### PR TITLE
Add implementation for hot reload of configuration with  same domain from multiple files.

### PR DESCRIPTION
**Pull request by anannaya**
Basically we enable the global ratelimit at the istio ingress layer , where the EnvoyFilter configuartion is applied with a generic domain name .  The current ratelimit service does not support updating/adding descriptors under same domain name.(Duplicate domain not supported) .  This implementation is about supporting the addition of descriptors existing domain. 

For Example:   Global gateway filter would be like below  with a generic domain name **vhost-ratelimit**
```
piVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: gateway-filter-ratelimit
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway-ratelimit
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: GATEWAY
        listener:
          filterChain:
            filter:
              name: "envoy.http_connection_manager"
              subFilter:
                name: "envoy.router"
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.rate_limit
          config:
            domain: vhost-ratelimit
            rate_limit_service:
              grpc_service:
                envoy_grpc:
                  cluster_name: rate_limit_service
                timeout: 0.25s
    - applyTo: CLUSTER
      match:
        cluster:
          service: ratelimit.ratelimit.svc.cluster.local
      patch:
        operation: ADD
        value:
          name: rate_limit_service
          type: STRICT_DNS
          connect_timeout: 0.25s
          lb_policy: ROUND_ROBIN
          http2_protocol_options: {}
          hosts:
            - socket_address:
                address: ratelimit.ratelimit.svc.cluster.local
                port_value: 8081

```
Next 
Basicaly I use a **kiwigrid/k8s-sidecar**. to reload the ratelimit  configuration from configmaps which is created by multiple applications. Sample Configmap looks like below , And which also has the Virtualservice specific envoyfilter which takes care of ratelimting the application requests.
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: configreload1
  labels:
    app: ratelimit
    ratelimit_config: "yes"
  namespace: pf
data:
  # statsd-mapping file
  configreload.yaml: |
    domain: vhost-ratelimit
    descriptors:
      - key: plann
        value: BASICCCCC
        descriptors:
          - key: accountttttt
            rate_limit:
              unit: minute
              requests_per_unit: 6
      - key: PATH
        value: "/productpage"
        rate_limit:
          unit: hour
          requests_per_unit: 6
```

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: filter-ratelimit-bookinfo
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
    - applyTo: VIRTUAL_HOST
      match:
        context: GATEWAY
        routeConfiguration:
          vhost:
            name: "bookinfo.xxxxxxx.com:443"
            route:
              action: ANY
      patch:
        operation: MERGE
        value:
          rate_limits:
            - actions:
                - request_headers:
                    header_name: ":path"
                    descriptor_key: "PATH"
```

### CODE REVIEW

**Comment by anannaya** [comment]
@ysawa0  pls review

**Comment by ysawa0** [comment]
@anannaya apologizes I haven't had any time recently. I'll review as soon as I'm able.

File: README.md [hunk 1]
```diff
@@ -345,6 +345,7 @@ RUNTIME_IGNOREDOTFILES default:"false"
 There are two methods for triggering a configuration reload:
 1. Symlink RUNTIME_ROOT to a different directory.
 2. Update the contents inside `RUNTIME_ROOT/RUNTIME_SUBDIRECTORY/config/` directly.
+3. Add/Update the existing domain descriptors `UPSERT_DESCRIPTORS` default `false`
```
Comment (mattklein123) [inline README.md:348]: What does `UPSERT_DESCRIPTORS` mean? Can you describe what this does and/or have a better name?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks. Can you provide more context and docs and also merge main and then we can take a look? Thanks.

/wait

File: README.md (hunk 1 above)
Comment (anannaya) [reply README.md:348]: @mattklein123  Could you suggest a better name?  I am actually looking for feature where apps can add/delete descriptors dynamically .

Comment (anannaya) [reply README.md:348]: @mattklein123  integration tests are failing , I think its from latest merge. Could you help me in resolving that ?

Comment (mattklein123) [reply README.md:348]: Main branch is broken right now, see https://github.com/envoyproxy/ratelimit/pull/213.

I still don't understand what this change does. Can you update the description to remove Envoy/Istio and just be clear about what we are trying to do with the rate limit service itself? Thank you.

/wait

Comment (anannaya) [reply README.md:348]: @mattklein123  Added few lines , I basically want to update the existing domain descriptiors using config files.(RUNTIME_WATCH_ROOT = false)

Comment (mattklein123) [reply README.md:348]: Sorry what lines did you add? Please fully update the description to remove Istio/Envoy and talk about this change in isolation in terms of what it does for the ratelimit service. Thank you!

/wait

**Comment by MischaWeerwag** [comment]
This is still an issue, bump.

**Comment by m-rcl** [comment]
Still an issue I'm encountering right now but the PR was closed. 
Any suggestions how we should proceed from here? 
I could offer to raise a new PR or build upon the existing one from @anannaya if there are no other implications of this proposed change, that I'm not aware of.

**Comment by anannaya** [comment]
I am using from private branch with some other enhancement, would be happy if you could raise a PR @m-rcl . Watch over envoy new project envoy-gateway

**Pull request closed without merging**